File: api/src/main/java/com/grash/model/SafetyInstruction.java
package com.grash.model;

import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.grash.model.abstracts.DateAudit;
import com.grash.model.enums.InstructionType;
import lombok.Data;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import java.time.LocalDateTime;

@Entity
@Data
@NoArgsConstructor
public class SafetyInstruction extends DateAudit {
    @Id
    @GeneratedValue(strategy = GenerationType.AUTO)
    private Long id;
    
    @ManyToOne
    @JsonIgnoreProperties({"company", "workers", "safetyInstructions", "assets", "parts"})
    private Location location; // Standortspezifische Unterweisung
    
    @ManyToOne
    @JsonIgnoreProperties({"employees", "safetyInstructions", "calendarEntries", "assets", "locations", "parts"})
    private Vendor vendor; // Zugehöriger Auftragnehmer
    
    @Column(length = 500)
    private String title;
    
    @Column(length = 2000)
    private String description;
    
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    private LocalDateTime instructionDate;
    
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    private LocalDateTime expirationDate; // Standard: 12 Monate ab Unterweisungsdatum
    
    @Enumerated(EnumType.STRING)
    @Column(length = 50)
    private InstructionType type; // VIDEO, DOCUMENT, LINK
    
    @Column(length = 1000)
    private String instructionMaterialUrl; // URL zu Dokument/Video
    
    @Column(length = 255)
    private String instructionMaterialFileId; // Referenz auf hochgeladenes File
    
    @ManyToOne
    @JsonIgnoreProperties({"role", "company", "teams", "password", "workers"})
    private OwnUser instructor; // Durchführender
    
    @ManyToOne
    @JsonIgnoreProperties({"vendor", "customer", "currentSafetyInstruction", "createdBy", "updatedBy"})
    private ContractorEmployee employee; // Unterwiesener Mitarbeiter
    
    private boolean completed;
    
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd'T'HH:mm:ss.SSS'Z'")
    private LocalDateTime completionDate;
    
    @Column(length = 10000)
    private String signatureData; // Elektronische Signatur
    
    @Column(length = 255)
    private String signatureName; // Name des Unterzeichners
    
    @ManyToOne
    @JsonIgnoreProperties({"role", "company", "teams", "password", "workers"})
    private OwnUser createdBy; // User who created this instruction
    
    // Transient fields for JSON serialization (computed dynamically, not persisted in DB)
    @Transient
    public Long getLocationId() {
        return location != null ? location.getId() : null;
    }
    
    @Transient
    public Long getVendorId() {
        return vendor != null ? vendor.getId() : null;
    }
    
    @Transient
    public Long getInstructorId() {
        return instructor != null ? instructor.getId() : null;
    }
    
    @Transient
    public Long getEmployeeId() {
        return employee != null ? employee.getId() : null;
    }
}

File: frontend/src/models/owns/safetyInstruction.ts
import { Audit } from './audit';

import Location from './location';
import { Vendor } from './vendor';
import { OwnUser } from '../user';
import { ContractorEmployee } from './contractorEmployee';

export interface SafetyInstruction extends Audit {
  id: number;
  title: string;
  description: string;
  instructionDate: string;
  expirationDate: string;
  type: 'VIDEO' | 'DOCUMENT' | 'LINK';
  instructionMaterialUrl: string;
  instructionMaterialFileId: string;
  location?: Location | null;
  vendor?: Vendor | null;
  instructor?: OwnUser | null;
  employee?: ContractorEmployee | null;
  completed: boolean;
  completionDate: string;
  signatureData: string;
  signatureName: string;
  // Keep ID fields for form compatibility
  locationId?: number;
  vendorId?: number | null;
  instructorId?: number;
  employeeId?: number;
}

export interface SafetyInstructionMiniDTO {
  id: number;
  title: string;
  completed: boolean;
  expirationDate: string;
}

export const safetyInstructions: SafetyInstruction[] = [
  {
    id: 1,
    title: 'Sicherheitsunterweisung für Standort A',
    description: 'Grundlegende Sicherheitsvorschriften',
    instructionDate: '2023-01-15T09:00:00',
    expirationDate: '2024-01-15T09:00:00',
    type: 'DOCUMENT',
    instructionMaterialUrl: 'https://example.com/safety.pdf',
    instructionMaterialFileId: 'file123',
    locationId: 1,
    vendorId: 1,
    instructorId: 1,
    employeeId: 1,
    completed: true,
    completionDate: '2023-01-15T10:30:00',
    signatureData: 'base64signature',
    signatureName: 'Max Mustermann',
    createdBy: 1,
    updatedBy: 1,
    createdAt: '2023-01-10T08:00:00',
    updatedAt: '2023-01-15T10:30:00'
  }
];